### ISSUE TITLE
adding indexed column to `DynamicTable` as list of lists

### ISSUE THREAD

**Issue by bendichter** [NeurodataWithoutBorders/pynwb#833]
## 2) Feature Request
If you are requesting a feature please provide the following:

### Problem/Use Case
Currently it is easy to add an indexed column and then add data row-by-row, but you cannot add an indexed column without first supplying the index. I think it would be more convenient to have the option to enter a list of lists and have the API automatically create the index.

for example, `units.add_column('col_name', 'col_description', [[1, 2, 3],[1, 2]], index=True)`

gives the error:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-169838992b6c> in <module>
      3 with NWBHDF5IO(file,'a') as io:
      4     nwb = io.read()
----> 5     nwb.units.add_column('col_name', 'col_description', [[1, 2, 3],]*len(nwb.units), index=True)
      6 
      7     position_object = nwb.modules['behavior']['EightMazePosition_position']['EightMazePosition_norm_spatial_series']

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    379                         raise_from(ExceptionType(msg), None)
    380 
--> 381                 return func(self, **parsed['args'])
    382         else:
    383             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/core.py in add_column(self, **kwargs)
   1163             elif isinstance(index, bool):        # make empty VectorIndex
   1164                 if len(col) > 0:
-> 1165                     raise ValueError("cannot pass empty index with non-empty data to index")
   1166                 col_index = VectorIndex(name + "_index", list(), col)
   1167             else:                                # make VectorIndex with supplied data

ValueError: cannot pass empty index with non-empty data to index
```

## Checklist

- [x] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?

**Comment by lvsltz** [comment]
Just to make sure, is this still the current behaviour, i.e. data in indexed columns has to be added iteratively?

**Comment by bendichter** [comment]
@lvsltz I believe you can use `index=VectorIndex(data=[...], ...)` instead of `index=True`

**Comment by lvsltz** [comment]
Thanks @bendichter for the tip! I'll give it a try

**Comment by luiztauffer** [comment]
This is still an Issue on `pynwb==2.0.1`, it can be reproduced with:
```python
nwbfile = NWBFile(
    session_description="desc",
    identifier="id123",
    session_start_time = datetime(2022, 1, 1, 0, 0, 0, tzinfo=tz.gettz("US/Pacific")),
)

nwbfile.add_trial(start_time=1.0, stop_time=3.0)
nwbfile.add_trial(start_time=3.0, stop_time=10.0)

nwbfile.add_trial_column(
    name="var2_times",
    description="desc",
    data=[[1,2],[3,5,6,7]],
    index=True,
)
```

error:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [34], in <cell line: 10>()
      7 nwbfile.add_trial(start_time=1.0, stop_time=3.0)
      8 nwbfile.add_trial(start_time=3.0, stop_time=10.0)
---> 10 nwbfile.add_trial_column(
     11     name="var2_times",
     12     description="desc",
     13     data=[[1,2],[3,5,6,7]],
     14     index=True,
     15 )

File ~/anaconda3/envs/env_oconnor/lib/python3.8/site-packages/hdmf/utils.py:583, in docval.<locals>.dec.<locals>.func_call(*args, **kwargs)
    581 def func_call(*args, **kwargs):
    582     pargs = _check_args(args, kwargs)
--> 583     return func(args[0], **pargs)

File ~/anaconda3/envs/env_oconnor/lib/python3.8/site-packages/pynwb/file.py:672, in NWBFile.add_trial_column(self, **kwargs)
    667 """
    668 Add a column to the trial table.
    669 See :py:meth:`~hdmf.common.DynamicTable.add_column` for more details
    670 """
    671 self.__check_trials()
--> 672 call_docval_func(self.trials.add_column, kwargs)

File ~/anaconda3/envs/env_oconnor/lib/python3.8/site-packages/hdmf/utils.py:424, in call_docval_func(func, kwargs)
    422 def call_docval_func(func, kwargs):
    423     fargs, fkwargs = fmt_docval_args(func, kwargs)
--> 424     return func(*fargs, **fkwargs)

File ~/anaconda3/envs/env_oconnor/lib/python3.8/site-packages/hdmf/utils.py:583, in docval.<locals>.dec.<locals>.func_call(*args, **kwargs)
    581 def func_call(*args, **kwargs):
    582     pargs = _check_args(args, kwargs)
--> 583     return func(args[0], **pargs)

File ~/anaconda3/envs/env_oconnor/lib/python3.8/site-packages/hdmf/common/table.py:736, in DynamicTable.add_column(self, **kwargs)
    734 elif isinstance(index, bool):  # make empty VectorIndex
    735     if len(col) > 0:
--> 736         raise ValueError("cannot pass empty index with non-empty data to index")
    737     col_index = VectorIndex(name + "_index", list(), col)
    738     self.__add_column_index_helper(col_index)

ValueError: cannot pass empty index with non-empty data to index
```

**Comment by oruebel** [comment]
This feature request would need to be addressed in HDMF since this is ultimately a feature of ``DynamicTable.add_column`` rather than a feature that is specific for ``TimeIntervals``. 

**Approach 1:**
I've implemented a possible solution in the following PR: https://github.com/hdmf-dev/hdmf/pull/728. Here, I've implemented this as functionality of ``DynamicTable.add_column`` (i.e., we have the API do "magic" under the hood), so that it would look for the user as follows:

```
nwbfile.add_trial_column(
    name="var2_times",
    description="desc",
    data=[[1,2],[3,5,6,7]],
    index=True,
)
```

The approach also works for multiple levels of indexes, i.e., data could also be a list of lists of lists:

```
nwbfile.add_trial_column(
    name="var3_times_times",
    description="desc",
    data=[[[1, 1], [2,]],[[3, 3, 3], [5,], [6,], [7,]]],
    index=2,
)
```

**Approach 2:**
An alternative approach would be to provide helper functions to support: a) flattening of nested lists and b) computation of the index data for a nested list. From a user perspective this would then look something like:

```
data = [[1,2],[3,5,6,7]]
nwbfile.add_trial_column(
    name="var2_times",
    description="desc",
    data=flatten_nested_list(data),
    index=index_from_nested_list(data)
)
```

**Advantages/Disadvantages**  This approach would 
**+** make it explicit that there is overhead for flattening the nested lists and computing the index 
**+** avoid adding more complexity to ``DynamicTable.add_column``
**-** a user needs to make 2 extra function calls, which may be less intuitive

@rly @bendichter  thoughts
